### PR TITLE
Implement Changes for embedded samples and Analysis for genparticle studies

### PR DESCRIPTION
**Pull request by harrypuuter**
In this PR, several changes are included:

- additional lepton veto in mm channel
- changes needed for embedded samples
- a config for studies on genlevel

### CODE REVIEW

File: code_generation/quantities/output.py [hunk 1]
```diff
@@ -7,6 +7,8 @@
 base_muons_mask = Quantity("base_muons_mask")
 good_muons_mask = Quantity("good_muons_mask")
 veto_muons_mask = Quantity("veto_muons_mask")
+veto_muons_mask_2 = Quantity("veto_muons_mask_2")
+veto_muons_mask_combined = Quantity("veto_muons_mask_combined")
```
Comment (nshadskiy) [inline code_generation/quantities/output.py:11]: Where is `veto_muons_mask_combined` used?

Comment (harrypuuter) [reply code_generation/quantities/output.py:11]: upps, a remnant :wink:

File: src/pairselection.hxx [hunk 2]
```diff
@@ -48,15 +78,159 @@ auto buildgenpair(auto &df, const std::string &recopair,
...
+            // in the case of two different daughter particles, we need find
+            // each one seperately
+            for (const auto &gen_candidate_1 : gen_candidates_1) {
+                if (check_mother(genparticles, gen_candidate_1.index,
+                                 mother_pdgid)) {
+                    genpair[0] = gen_candidate_1.index;
+                    break;
+                }
+            }
+            for (const auto &gen_candidate_2 : gen_candidates_1) {
```
Comment (nshadskiy) [inline src/pairselection.hxx:198]: You want to loop over `gen_candidates_2` here, I think.

File: src/pairselection.hxx [hunk 3]
```diff
@@ -48,15 +78,159 @@ auto buildgenpair(auto &df, const std::string &recopair,
...
+/// pairselection::mutau::PairSelectionAlgo) returns indices, that are
+/// not -1. Events, where any of the particle indices is -1 are vetoed
+/// by this filter.
 ///
 /// \param df The input dataframe
 /// \param flagname The name of the generated flag column
-/// \param pairname The name of the column, containing the particle indices
-/// index of the particle in the particle quantity vectors.
+/// \param pairname The name of the column, containing the particle
+/// indices index of the particle in the particle quantity vectors.
```
Comment (nshadskiy) [inline src/pairselection.hxx:233]: I think this should be something like this.
```
\param pairname The name of the column, containing the 
indices of the particles in the particle quantity vectors.
```

File: src/pairselection.hxx [hunk 4]
```diff
@@ -288,15 +475,119 @@ auto PairSelectionAlgo() {
...
+        const auto good_phis =
+            ROOT::VecOps::Take(muonphi, original_muon_indices);
+        const auto good_masses =
+            ROOT::VecOps::Take(muonmass, original_muon_indices);
+        auto fourVecs = ROOT::VecOps::Construct<ROOT::Math::PtEtaPhiMVector>(
+            good_pts, good_etas, good_phis, good_masses);
+        float mass_difference = -1.0;
+        float zmass_candidate = -1.0;
+        auto selected_mu_indices = std::vector<int>{};
+        if (original_muon_indices.size() > 2) {
```
Comment (nshadskiy) [inline src/pairselection.hxx:514]: `>= 2`?

**Review by nshadskiy — APPROVED** [review]
Looks nice! But check the comments before merging.

File: src/pairselection.hxx (hunk 3 above)
Comment (harrypuuter) [reply src/pairselection.hxx]: changes

**Pull request merged**